test/test_common/simulated_time_system.cc

```diff
@@ -0,0 +1,194 @@
+#include "test/test_common/simulated_time_system.h"
+
+#include <chrono>
+
+#include "envoy/event/dispatcher.h"
+
+#include "common/common/assert.h"
+#include "common/common/lock_guard.h"
+#include "common/event/real_time_system.h"
+#include "common/event/timer_impl.h"
+
+#include "event2/event.h"
+
+namespace Envoy {
+namespace Event {
+
+// Our simulated alarm inherits from TimerImpl so that the same dispatching
+// mechanism used in RealTimeSystem timers is employed for simulated alarms.
+// using libevent's event_active(). Note that libevent is placed into
+// thread-safe mode due to the call to evthread_use_pthreads() in
+// source/common/event/libevent.cc.
+class SimulatedTimeSystem::Alarm : public TimerImpl {
+public:
+  Alarm(SimulatedTimeSystem& time_system, Libevent::BasePtr& libevent, TimerCb cb)
+      : TimerImpl(libevent, cb), time_system_(time_system), index_(time_system.nextIndex()),
+        armed_(false) {}
+
+  virtual ~Alarm();
+
+  // Timer
+  void disableTimer() override;
+  void enableTimer(const std::chrono::milliseconds& duration) override;
+
+  void setTime(MonotonicTime time) { time_ = time; }
+
+  /**
+   * Activates the timer so it will be run the next time the libevent loop is run,
+   * typiically via Dispatcher::run().
+   */
+  void activate() {
+    armed_ = false;
+    std::chrono::milliseconds duration = std::chrono::milliseconds::zero();
+    TimerImpl::enableTimer(duration);
+  }
+
+  MonotonicTime time() const {
+    ASSERT(armed_);
+    return time_;
+  }
+
+  uint64_t index() const { return index_; }
+
+private:
+  SimulatedTimeSystem& time_system_;
+  MonotonicTime time_;
+  uint64_t index_;
+  bool armed_;
+};
+
+// Compare two alarms, based on wakeup time and insertion order. Returns true if
+// a comes before b.
+
+// like strcmp (<0 for a < b, >0 for a > b), based on wakeup time and index.
+bool SimulatedTimeSystem::CompareAlarms::operator()(const Alarm* a, const Alarm* b) const {
+  if (a != b) {
+    if (a->time() < b->time()) {
+      return true;
+    } else if (a->time() == b->time() && a->index() < b->index()) {
+      return true;
+    }
+  }
+  return false;
+};
+
+// Each scheduler maintains its own timer
```

Comment:
nit: time system instead of timer?

Reply:
The TimeSystem is common across all the schedulers. At one point I thought each Scheduler would maintain its own set of timers, but it turns out to be easier to have the TimeSystem own the timers for every scheduler. The scheduler just holds onto the libevent_ base ptr, to allow activation when time moves forward. I'll add more commenting.

Reworded as:
```
// Each timer is maintained and ordered by a common TimeSystem, but is
// associated with a scheduler.  The scheduler associates the timers it creates
// with a libevent context, so that the timer callbacks are executed in via
// Dispatcher::run() in the proper thread.
```